docs/control-flow.md

```diff
@@ -0,0 +1,361 @@
+```
+
++++ {"id": "lIYdn1woOS1n"}
+
+`jax.lax` provides two other functions that allow branching on dynamic predicates:
+
+- [`lax.select`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.select.html) is
+  like a batched version of `lax.cond`, with the choices expressed as pre-computed arrays
+  rather than as functions.
+- [`lax.switch`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.switch.html) is
+  like `lax.cond`, but allows switching between any number of callable choices.
+
+In addition, `jax.numpy` provides several numpy-style interfaces to these functions:
+
+- [`jnp.where`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.where.html) with
+  three arguments is the numpy-style wrapper of `lax.select`.
+- [`jnp.piecewise`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.piecewise.html)
+  is a numpy-style wrapper of `lax.switch`, but switches on a list of boolean conditions rather than a single scalar index.
+- [`jnp.select`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.select.html) has
+  an API similar to `jnp.piecewise`, but the choices are given as pre-computed arrays rather
+  than as functions. It is implemented in terms of multiple calls to `lax.select`.
+
++++ {"id": "xkOFAw24eOMg"}
+
+### `while_loop`
+
+python equivalent:
+```
+def while_loop(cond_fun, body_fun, init_val):
+  val = init_val
+  while cond_fun(val):
+    val = body_fun(val)
+  return val
+```
+
+```{code-cell}
+:id: jM-D39a-c436
+:outputId: 552fe42f-4d32-4e25-c8c2-b951160a3f4e
+
+init_val = 0
+cond_fun = lambda x: x < 10
+body_fun = lambda x: x+1
+lax.while_loop(cond_fun, body_fun, init_val)
+# --> array(10, dtype=int32)
+```
+
++++ {"id": "apo3n3HAeQY_"}
+
+### `fori_loop`
+python equivalent:
+```
+def fori_loop(start, stop, body_fun, init_val):
+  val = init_val
+  for i in range(start, stop):
+    val = body_fun(i, val)
+  return val
+```
+
+```{code-cell}
+:id: dt3tUpOmeR8u
+:outputId: 7819ca7c-1433-4d85-b542-f6159b0e8380
+
+init_val = 0
+start = 0
+stop = 10
+body_fun = lambda i,x: x+i
+lax.fori_loop(start, stop, body_fun, init_val)
+# --> array(45, dtype=int32)
+```
+
++++ {"id": "SipXS5qiqk8e"}
+
+### Summary
+
+$$
+\begin{array} {r|rr}
+\hline \
+\textrm{construct}
+& \textrm{jit}
+& \textrm{grad} \\
+\hline \
+\textrm{if} & ❌ & ✔ \\
+\textrm{for} & ✔* & ✔\\
+\textrm{while} & ✔* & ✔\\
+\textrm{lax.cond} & ✔ & ✔\\
+\textrm{lax.while_loop} & ✔ & \textrm{fwd}\\
+\textrm{lax.fori_loop} & ✔ & \textrm{fwd}\\
+\textrm{lax.scan} & ✔ & ✔\\
+\hline
+\end{array}
+$$
+
+<center>
+
+$\ast$ = argument-<b>value</b>-independent loop condition - unrolls the loop
+
+</center>
+
+## Logical operators
+
+`jax.numpy` provides `logical_and`, `logical_or`, and `logical_not`, which operate element-wise on arrays and can be evaluated under `jit` without recompiling. Like their Numpy counterparts, the binary operators do not short circuit. Bitwise operators (`&`, `|`, `~`) can also be used with `jit`.
```

Comment:
An example of this would be useful, but I'm happy to get this in and add an example in a followup PR!

Reply:
Sounds good, I'll add that as another task on #24632 .